### PR TITLE
raw events

### PR DESCRIPTION
**Pull request by tilsche**
for #64

### CODE REVIEW

File: src/perf/event_provider.cpp [hunk 1]
```diff
@@ -475,12 +488,23 @@ EventProvider::EventProvider()
 
 const CounterDescription& EventProvider::cache_event(const std::string& name)
 {
+    // Format for raw events is rNNNN
+    static const std::regex raw_regex("r[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]");
```
Comment (tilsche) [inline src/perf/event_provider.cpp]: Please use `[[:xdigit:]]{1-16}`.

File: src/perf/event_provider.cpp [hunk 2]
```diff
@@ -137,7 +138,7 @@ constexpr std::uint64_t bit(int bitnumber)
 {
     return static_cast<std::uint64_t>(1_u64 << bitnumber);
 }
-}
+} // namespace
```
Comment (tilsche) [inline src/perf/event_provider.cpp]: :-)

File: src/perf/event_provider.cpp [hunk 3]
```diff
@@ -345,6 +346,18 @@ static void event_description_update(CounterDescription& event, std::uint64_t va
...
 }
 
+const CounterDescription raw_read_event(const std::string& ev_desc)
+{
+    uint64_t code = std::stoull(ev_desc.substr(1), nullptr, 16);
+    const CounterDescription event(ev_desc, PERF_TYPE_RAW, code, 0);
+
+    if (!event_is_openable(event))
+    {
+        throw EventProvider::InvalidEvent("cannot open requested event");
```
Comment (bmario) [inline src/perf/event_provider.cpp]: Add the name of the event, which failed

Comment (cvonelm) [reply src/perf/event_provider.cpp]: Is that needed? The error message in that case ( Example: " 'r003c' does not name a known event, ignoring! (reason: cannot open requested event)") already contains the name of the event

Comment (bmario) [reply src/perf/event_provider.cpp]: Oh, it get's added somewhere higher in call stack?

Comment (cvonelm) [reply src/perf/event_provider.cpp]: Seems like that

**Pull request merged**